compendium/DeclarativeServices/src/ComponentContextImpl.cpp

```diff
@@ -103,6 +106,133 @@ ComponentContextImpl::GetProperties() const
   return configManagerPtr->GetProperties();
 }
 
+/**
+ * Returns a service interface pointer from a given InterfaceMap instance.
+ * 
+ * This function uses a reference parameter (boolean) to determine whether or not
+ * a nullptr returned from ExtractInterface is due to a construction/activation
+ * failure or because the information passed to the function is garbage.
+ * 
+ * This function has a similar implementation to ExtractService except it uses
+ * a boolean to keep track of whether or not the returned service interface
+ * pointer is null due to an error during construction or activation
+ *
+ * - map, a InterfaceMap instance.
+ * - interfaceId, The interface id string.
+ * - isSIPointerFoundOrMapEmpty, a boolean which represents if the return value was in the map
+ *     or the map is nullptr
+ * - return value: The service interface pointer for the service interface id or
+ *                 nullptr if "map" does not contain an entry for the given type.
+ */
+static std::shared_ptr<void> GetInterfacePointerAndCheckError(
+  const InterfaceMapConstPtr& map,
+  const std::string& interfaceId,
+  bool& isSIPointerFoundOrMapEmpty)
```

Comment:
I recommend following the cpp core guidelines as it relates to return values and "out" parameters, specifically: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rf-out and https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rf-out-multi.